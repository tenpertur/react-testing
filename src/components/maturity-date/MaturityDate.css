div.maturity-date {
    display: inline;
}