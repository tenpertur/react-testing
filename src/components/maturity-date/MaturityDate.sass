p
  &.maturity-date-value
    display: inline
    color: green
    padding: 0
    margin: 0

    &.error
      color: red

div
  &.maturity-date
    display: block
    padding-top: 1em

