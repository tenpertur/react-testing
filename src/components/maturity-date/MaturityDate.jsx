import React, {useEffect, useState} from "react";
import './MaturityDate.sass';
import {Skeleton} from "Components/skeleton/Skeleton";
import {defer, EMPTY, from, Subject, timer} from "rxjs";
import {catchError, debounceTime, map, mergeMap, tap} from "rxjs/operators";
import axios from "axios";

export const useMaturityDate = () => {
    const [startDate, setStartDate] = useState();
    const [duration, setDuration] = useState();
    const [maturityDate, setMaturityDate] = useState();
    const [isLoading, setIsLoading] = useState(false);
    const [isError, setIsError] = useState(false);
    // const [subject$] = useState(
    //     new Subject()
    // );
    // const [stream$] = useState(
    //     subject$.asObservable().pipe(
    //         debounceTime(300),
    //         mergeMap(data => {
    //             return defer(() => {
    //                 const progress = timer(300).subscribe(_ => {
    //                     setIsLoading(true);
    //                 });
    //                 return from(
    //                     axios.get(`http://localhost:8080/date?duration=${data.duration}&from=${data.startDate}`)
    //                 ).pipe(
    //                     tap(
    //                         _ => {
    //                             progress.unsubscribe();
    //                         },
    //                         _ => {
    //                             progress.unsubscribe();
    //                         },
    //                         _ => {
    //                             progress.unsubscribe();
    //                         }
    //                     ),
    //                     map(response => {
    //                         setIsLoading(false);
    //                         const error = response.status !== 200;
    //                         setIsError(error);
    //                         return response.text();
    //                     }),
    //                     catchError(_ => {
    //                         setIsLoading(false);
    //                         setIsError(true);
    //                         return EMPTY;
    //                     })
    //                 );
    //             });
    //         })
    //     )
    // );
    // useEffect(() => {
    //     const subscription = stream$.subscribe(
    //         next => {
    //             setMaturityDate(next);
    //             setIsLoading(false);
    //         },
    //         _ => {
    //             setIsError(true);
    //             setIsLoading(false);
    //         }
    //     );
    //     return () => {
    //         setIsLoading(false);
    //         subscription.unsubscribe();
    //     };
    // }, []);
    //
    useEffect(() => {
        if (startDate && duration) {
            subject$.next({duration: duration, startDate: startDate});
        }
    }, []);

    useEffect(() => {
        if (!startDate || !duration) {
            return;
        }
        setIsLoading(true)
        fetch(`http://localhost:8080/date?duration=${duration}&from=${startDate}`)
            .then(resp => {
                const error = resp.status !== 200
                setIsError(error)
                setIsLoading(false)
                return resp.text()
            }).then(text => {
            setMaturityDate(text)
        })
            .catch(_ => {
                setIsError(true)
                setIsLoading(false)
            })
    }, [startDate, duration])
    return [maturityDate, isLoading, isError, setStartDate, setDuration]
}

export const MaturityDate = ({maturityDate, isLoading, isError}) => {
    let element;
    if (isError) {
        element = <p className="maturity-date-value error">Error during calculation</p>;
    } else if (isLoading) {
        element = <Skeleton height={"1em"} width={"14em"}/>
    } else {
        element = <p className="maturity-date-value">Maturity date: {maturityDate}</p>
    }
    return <div className="maturity-date">
        {element}
    </div>
}