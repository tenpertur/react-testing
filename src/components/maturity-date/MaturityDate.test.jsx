import React from 'react';
import renderer from 'react-test-renderer';
import {MaturityDate} from "Components/maturity-date/MaturityDate";

describe("MaturityDate tests", () => {

    it('renders correctly maturity date', () => {
        const component = <MaturityDate maturityDate="2021-10-11" isLoading={false} isError={false}/>
        const tree = renderer
            .create(component)
            .toJSON()
        expect(tree).toMatchSnapshot();
    })
    it('renders maturity date skeleton when loading', () => {
        const component = <MaturityDate maturityDate="2021-10-11" isLoading={true} isError={false}/>
        const tree = renderer
            .create(component)
            .toJSON()
        expect(tree).toMatchSnapshot();
    })

    it('renders error when error', () => {
        const component = <MaturityDate maturityDate="2021-10-11" isLoading={false} isError={true}/>
        const tree = renderer
            .create(component)
            .toJSON()
        expect(tree).toMatchSnapshot();
    })

})

