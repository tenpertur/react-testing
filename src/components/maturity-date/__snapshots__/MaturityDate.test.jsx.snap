// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaturityDate tests renders correctly maturity date 1`] = `
<div
  className="maturity-date"
>
  <p
    className="maturity-date-value"
  >
    Maturity date: 
    2021-10-11
  </p>
</div>
`;

exports[`MaturityDate tests renders error when error 1`] = `
<div
  className="maturity-date"
>
  <p
    className="maturity-date-value error"
  >
    Error during calculation
  </p>
</div>
`;

exports[`MaturityDate tests renders maturity date skeleton when loading 1`] = `
<div
  className="maturity-date"
>
  <div
    className="skeleton shine"
    style={
      Object {
        "height": "1em",
        "width": "14em",
      }
    }
  />
</div>
`;
